A43-sdk-authorization.md

```diff
@@ -0,0 +1,707 @@
+} catch (IllegalArgumentException iae) {
+  // ...
+} catch (IOException ioe) {
+  // ...
+}
+Closeable closeable = authzServerInterceptor.scheduleRefreshes(
+    1, TimeUnit.HOURS, scheduledExecutor);
+Server server =
+    Grpc.newServerBuilderForPort(port, serverCreds)
+        .addService(service)
+        .intercept(authzServerInterceptor)
+        .build()
+        .start();
+// In the end, free up the resources used for policy refresh.
+closeable.close();
+```
+
+## Rationale
+
+gRPC Authorization internally implements RBAC Engine(s) based on Envoy [RBAC policy](https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/rbac/v3/rbac.proto).
+We decided to create a new policy language "SDK authorization policy" instead of 
+consuming Envoy RBAC directly due to following reasons:
+- Envoy RBAC is a complex language, and we preferred using a simple human readable 
+  policy language
+- With our own language, we can provide a stable API, even when Envoy undergoes 
+  versioning updates.
+
+Note that SDK authorization policy is a subset of Envoy RBAC, and it does not
+support all the fields that are present in Envoy RBAC. 
+
+RBAC policy provides service-level and method-level access control for a service.
+Engines process incoming RPC request attributes against policy configs and make a
+decision on whether to allow or deny the request. The decision depends on the type
+of policy (if the policy is an allowlist or denylist) and whether a matching policy
+was found.
+
+This engine implementation is shared by SDK-based authorization and xDS-based
+authorization. In xDS authorization, the xDS server sends an RBAC policy config to
+xDS-enabled gRPC server for authorization.([A41](A41-xds-rbac.md): xDS RBAC Support)
+
+Overall SDK authorization flow is as follows. User supplies gRPC SDK authorization
+policy to provider/interceptor. The provider then forwards the JSON policy to Policy
+translator. The translator converts JSON policy to Envoy RBAC protos (Allow and/or
+Deny policy). Translator errors out on I/O errors or if the policy does not
+represent JSON schema it currently supports. Ultimately the generated RBAC policies
+are used to create Envoy RBAC authorization engine(s). Then, for each incoming RPC
+request, we will invoke the Engines to get the authorization decision.
+
+SDK authorization can be enabled in xDS enabled servers, which means both
+authorization paths (SDK and xDS authorization) can co-exist. For a request to
+be allowed, both paths must allow the request.
+
+As mentioned previously, authorization APIs take policy in JSON format, instead of
+protobuf. This is done to avoid the dependency on protobuf. We have users in OSS
+that use gRPC without protobuf. Another reason is to have a consistent API across
+languages.
+
+## Implementation
+
+The implementation order will be C++, Go, Java and then wrapped languages. The 
+implementation will be done by ashithasantosh.
```

Comment:
Is this just "Done in all languages" now?

Reply:
Its done in C++ and Go. Java is currently under implementation.